### ISSUE TITLE
How to create a VM with disk and NIC rather than add after create ?

### ISSUE THREAD

**Issue by ruiruige** [vmware/pyvmomi-community-samples#324]
I carefully read some of the related samples.
It has `add_disk_to_vm.py` and `add_nic_to_vm.py`

However,  can I create a vm with a disk and a nic ?  

in `create_random_marvel_vms.py` it does not say.
It only says:
```
   vm_name = 'MARVEL-' + name
    datastore_path = '[' + datastore + '] ' + vm_name

    # bare minimum VM shell, no disks. Feel free to edit
    vmx_file = vim.vm.FileInfo(logDirectory=None,
                               snapshotDirectory=None,
                               suspendDirectory=None,
                               vmPathName=datastore_path)

    config = vim.vm.ConfigSpec(name=vm_name, memoryMB=128, numCPUs=1,
                               files=vmx_file, guestId='dosGuest',
                               version='vmx-07')
```

So, is there any method to create a VM with a disk?

**Comment by prziborowski** [comment]
You can do pretty much whatever is in the Reconfigure call in CreateVm itself.
I'd think you can combine what is in add_disk_to_vm.py and add_nic_to_vm.py to create_random_marval_vms.py and have it work (note I haven't looked at any of those files, but vim.vm.ConfigSpec is valid for reconfigure and createVm).

**Comment by ruiruige** [comment]
“”“
        unit_number = 0

        disk_spec = vim.vm.device.VirtualDeviceSpec()
        disk_spec.fileOperation = "create"
        disk_spec.operation = vim.vm.device.VirtualDeviceSpec.Operation.add
        disk_spec.device = vim.vm.device.VirtualDisk()
        
        disk_spec.device.backing = \
            vim.vm.device.VirtualDisk.FlatVer2BackingInfo()   # https://github.com/vmware/pyvmomi/tree/575ab56eb56f32f53c98f40b9b496c6219c161da/docs/vim/vm/device/VirtualDisk


        disk_spec.device.backing.thinProvisioned = True
        disk_spec.device.backing.diskMode = 'persistent'
        disk_spec.device.unitNumber = unit_number
        disk_spec.device.capacityInKB = size_gb * 1024 * 1024
        disk_spec.device.controllerKey = controller.key
”“”

“”“
        devices.append(disk_spec)


        config = vim.vm.ConfigSpec(name=vm_name, memoryMB=8196, numCPUs=4,
                                   files=vmx_file, guestId=guestId,
                                   version='vmx-11', deviceChange=devices) 
”“”

**Comment by lingfish** [comment]
I think it would be great if someone made this a complete, standalone example in this repository. I for one would appreciate it.

**Comment by taherbohari** [comment]
Yes @lingfish. I totally agree with your point. Having a full fledged example of create_vm with basic requirements in it like : 
- boot the vm using provided iso
- having a disk
- having a nic

**Comment by lingfish** [comment]
I copped out and made a "nude" VM with the bare minimums, and clone from that; seemed way too complicated to get SCSI controllers etc added.

**Comment by lingfish** [comment]
Why was this closed?

**Comment by ruiruige** [comment]
@lingfish 
really sorry, I did not notice that your need has not been finished,  I will reopen it now.

**Comment by uddyan92** [comment]
@all please someone help me to configure ip address
on th vm

**Comment by hitesh-smit** [comment]
I need to create a VM with hard disk, 3 NIC, 2 CD-ROM(with 2 iso attached), specific RAM and CPU(I got that from one of the samples). Need to do all it during creation. Can someone please help.

**Comment by rahulonmars** [comment]
> 

What was the error that you got ?

**Comment by wangzq96** [comment]
中文：根据create_vm.py进行增加硬盘
English：Add a hard drive according to create_vm.py
"""
def create_config_spec(
    datastore_name,
    name,
    memory=2048,
    guest="otherGuest",
    annotation="Sample",
    cpus=1,
):
    config = vim.vm.ConfigSpec()
    config.annotation = annotation
    config.memoryMB = int(memory)
    config.guestId = guest
    config.name = name
    config.numCPUs = cpus

    # List of all device controllers we plan to attach to this VM
    vmControllers = []

    # SCSI controller
    scsi_ctlr = vim.vm.device.VirtualDeviceSpec()
    scsi_ctlr.operation = vim.vm.device.VirtualDeviceSpec.Operation.add
    # scsi_ctlr.device = vim.vm.device.ParaVirtualSCSIController()
    scsi_ctlr.device = vim.vm.device.VirtualLsiLogicController()
    scsi_ctlr.device.deviceInfo = vim.Description()
    # scsi_ctlr.device.slotInfo = vim.vm.device.VirtualDevice.PciBusSlotInfo()
    # scsi_ctlr.device.slotInfo.pciSlotNumber = 16
    scsi_ctlr.device.controllerKey = 100
    scsi_ctlr.device.unitNumber = 3
    scsi_ctlr.device.busNumber = 0
    scsi_ctlr.device.hotAddRemove = True
    scsi_ctlr.device.sharedBus = "noSharing"
    scsi_ctlr.device.scsiCtlrUnitNumber = 7
    vmControllers.append(scsi_ctlr)

    # vDisk
    unit_number = 0
    controller = scsi_ctlr.device  # this is the controller we defined above
    disk_spec = vim.vm.device.VirtualDeviceSpec()
    disk_spec.fileOperation = "create"
    disk_spec.operation = vim.vm.device.VirtualDeviceSpec.Operation.add
    disk_spec.device = vim.vm.device.VirtualDisk()
    disk_spec.device.backing = vim.vm.device.VirtualDisk.FlatVer2BackingInfo()
    disk_spec.device.backing.diskMode = "persistent"
    disk_spec.device.backing.thinProvisioned = True
    disk_spec.device.backing.fileName = "[%s]%s.vmdk" % (datastore_name, name)
    disk_spec.device.unitNumber = unit_number
    disk_spec.device.capacityInKB = 10 * 1024 * 1024
    disk_spec.device.controllerKey = controller.key
    vmControllers.append(disk_spec)

    files = vim.vm.FileInfo()
    files.vmPathName = "[" + datastore_name + "]"
    config.files = files
    config.deviceChange = vmControllers
    return config
"""
中国码友可以关注，csdn,后续完整代码增加上去
![1704940738920](https://github.com/vmware/pyvmomi-community-samples/assets/61556782/03857b97-758b-4acd-9795-2537dc8455f1)
